source/index.lua

```diff
@@ -0,0 +1,215 @@
+	if options.level and (tonumber(options.level) == nil or options.level >= 0 or options.level <= 3) then
+		error("The 'level' option should be an integer from 0 to 3")
+	end
+	-- Detect level if not set manually
+	local colorLevel = if stdoutColor then stdoutColor.level else 0
+	object.level = if options.level == nil then colorLevel else options.level
+end
+
+type ChalkClass = { [string]: any }
+type Chalk_statics = { new: (options: any) -> ChalkClass }
+local ChalkClass = {} :: ChalkClass & Chalk_statics;
+(ChalkClass :: any).__index = ChalkClass
+function ChalkClass.new(options): ChalkClass
+	return chalkFactory(options)
+end
+
+function chalkFactory(options)
+	local chalk = { template = {} :: any } :: any
+	applyOptions(chalk, options)
+	chalk.template = function(...)
+		chalkTag(chalk.template, ...)
+	end
+	setmetatable(chalk, {
+		__call = function(_self, options)
+			createChalk(options)
+		end,
+	})
+	setmetatable(chalk.template, chalk)
+	setmetatable(chalk.template, {
+		__newindex = function(self, key, value)
+			if key == "new" then
+				error("'chalk.constructor()' is deprecated. Use 'new chalk.Instance()' instead.")
+			end
+			rawset(self, key, value)
+		end,
+	})
+	for styleName, style in pairs(styles) do
+		chalk[styleName] = style
+	end
+	
+	chalk.template.Instance = ChalkClass
+	return chalk.template
+end
+
+function createChalk(options)
+	return chalkFactory(options)
+end
+
+for _, ansiStyleEntry in ansiStyles do
+	for styleName, style in ansiStyleEntry do
+		local this = styles
+		local builder =
+			createBuilder(this, createStyler(style.open, style.close, this._styler), this._isEmpty)
+		this[styleName] = builder
+	end
+end
+
+styles.visible = (function()
+	local this = styles.visible
+	return createBuilder(this, this._styler, true)
+end)()
+
+local usedModels = { "rgb", "hex", "keyword", "hsl", "hsv", "hwb", "ansi", "ansi256" }
+
+for _, model in usedModels do
+	styles[model] = (function(...)
+		local this = styles[model]
+		local level = this.level
+		local styler =
+			createStyler(ansiStyles.color[levelMapping[level]][model](...), ansiStyles.color.close, this._styler)
+		return createBuilder(this, styler, this._isEmpty)
+	end)()
+
+	local bgModel = "bg" .. string.upper(string.sub(model, 1,1)) .. String.slice(model, 1)
+	styles[bgModel] = (function(...)
+		local this = styles[bgModel]
+		local level = this.level :: number
+		local styler =
+			createStyler(ansiStyles.bgColor[levelMapping[level]][model](...), ansiStyles.bgColor.close, this._styler)
+		return createBuilder(this, styler, this._isEmpty)
+	end)()
+end
+
+function createStyler(open, close, parent: any?)
+	local openAll
+	local closeAll
+	if parent == nil then
+		openAll = open
+		closeAll = close
+	else
+		openAll = parent.openAll .. open
+		closeAll = close .. parent.closeAll
+	end
+	return { open = open, close = close, openAll = openAll, closeAll = closeAll, parent = parent }
+end
+
+function createBuilder(self, _styler, _isEmpty)
```

Comment:
It looks like `_styler` and `_isEmpty` are used, should these be updated to remove the underscores? They look like they're supposed to be ignored parameters